azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -162,3 +162,82 @@ def __init__(self, sas_token, callback=None):
         """
         super(SetSasToken, self).__init__(callback=callback)
         self.sas_token = sas_token
+
+
+class SendIotRequestAndWaitForResponse(PipelineOperation):
+    """
+    A PipelineOperation object which wraps the common operation of sending a request to iothub with a request_id ($rid)
+    value and waiting for a response with the same $rid value.  This convention is used by both Twin and Provisioning
+    features.
+
+    Even though this is an base operation, it will most likely be generated and also handled by more specifics stages
+    (such as IotHub or Mqtt stages).
+
+    The type of the request payload and the response payload is undefined at this level.  The type of the payload is defined
+    based on the type of request that is being executed.  If types need to be converted, that is the responsibility of
+    the stage which creates this operation, and also the stage which executes on the operation.
+
+    :ivar status_code: The status code returned by the response.  Any value under 300 is considered success.
+    :type status_code: int
+    :ivar response_body: The body of the response.
+    :type response_payload: Undefined
+    """
+
+    def __init__(self, request_type, verb, resource, request_body, callback=None):
+        """
+        Initializer for SendIotRequestAndWaitForResponse objects
+
+        :param str request_type: The type of request.  This is a string which is used by transport-specific stages to
+          generate the actual request.  For example, if request_type is "twin", then the iothub_mqtt stage will convert
+          the request into an MQTT publish with topic that begins with $iothub/twin
+        :param str verb: The verb for the request, such as "POST", "GET", etc.
+        :param str resource: The resource that the verb is acting on.  For twin request with verb "GET", this is most
+          likely the string "/" which retrieves the entire twin resource
+        :param request_body: The body of the request.  This is a required field, and a single space can be used to denote
+          an empty body.
+        :type request_body: Undefined
```

Comment:
>Undefined [](start = 28, length = 9)

I thought it was a string?

Reply:
Everything that comes in over MQTT arrives as a byte array.  At this level, the code doesn't care what it is.  It's only the producer and the consumer that care.  I don't want to define it if I don't have to because i don't want to force multiple conversions if I don't have to. 

---
In reply to: [288342762](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288342762) [](ancestors = 288342762)